gapir/client/session.go

```diff
@@ -167,31 +189,50 @@ func (s *session) newADB(ctx context.Context, d adb.Device, abi *device.ABI) err
 	if !ok {
 		return log.Errf(ctx, nil, "Unsupported architecture: %v", abi.Architecture)
 	}
-	if err := d.Forward(ctx, localPort, adb.NamedAbstractSocket(socket)); err != nil {
-		return log.Err(ctx, err, "Forwarding port")
+	apkDir, err := apk.FileDir(ctx)
+	if err != nil {
+		return log.Errf(ctx, err, "Getting gapid.apk files directory")
 	}
-	s.onClose(func() { d.RemoveForward(ctx, localPort) })
+	// Wait for the socket file to be created
+	socketPath := strings.Join([]string{apkDir, socket}, "/")
+	err = task.Retry(ctx, maxCheckSocketFileAttempts, checkSocketFileRetryDelay,
+		func(ctx context.Context) (bool, error) {
+			str, err := d.Shell("run-as", apk.Name, "ls", socketPath).Call(ctx)
```

Comment:
Has this been tested on rooted / non rooted / non-google devices?

Reply:
Tested with Pixel C (rooted and unrooted), Pixel 2 (unrooted)

Reply:
Just tested with XiaoMi 5S (Android 7.0), it works! (Just need to enable all USB debugging options)